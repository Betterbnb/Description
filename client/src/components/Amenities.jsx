import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import AmenityModal from './AmenityModal.jsx';

const StyledAmenities = styled.div`
  font-family: Circular, Helvetica Neue, Helvetica, Arial, sans-serif;
  color: #484848;
  margin: 20px 20px 10px 10px;
  font-size: 15px;
  font-weight: 320;
  line-height: 1.4;
  letter-spacing: 0.5px;
`;

const AmenitiesContainer = styled.div`
  display: grid;
  grid-template-columns: repeat(2, 1fr);
`;

const AmenityContainer = styled.div`
  display: flex;
  margin: 4px 4px 4px -2px;
`;

const AmenityIcon = styled.img`
  height: 32px;
  width: 32px;
  margin: -5px 5px 5px -5px;
`;

const ExpansionLink = styled.button`
  font-size: 16px;
  border: none;
  outline: none;
  font-weight: 500;
  color: #00abb2;
  padding: 0;
  margin-top: 10px;
`;

const imageUrls = {
  'essentials (towels, bedsheets, soap, toilet paper, pillows)': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/essentials.png',
  'wifi': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/wifi.png',
  'shampoo': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/shampoo.png',
  'TV': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/TV.png',
  'desk/workspace': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/laptopWorkspace.png',
  'fireplace': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/fireplace.png',
  'parking': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/parking.png',
  'gym': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/gym.png',
  'pool': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/pool.png',
  'dining': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/dining.png',
  'hottub': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/hottub.png',
  'washer': 'https://s3-us-west-1.amazonaws.com/betterbnb-description/washer.png',
};

const title = {
  fontSize: '16px',
  fontWeight: '500',
  marginBottom: '15px'
};

const capitalize = (word) => {
  return word.slice(0, 1).toUpperCase() + word.slice(1);
}

const assembleIcons = ({ basics, facilities, dining, safety }) => {
  let icons = [];
  let allAmenities = basics.concat(facilities, dining, safety);
  for (let i = 0; i < allAmenities.length; i += 1) {
    if (allAmenities[i] in imageUrls) {
      icons.push([allAmenities[i], imageUrls[allAmenities[i]]]);
    }
    if (icons.length >= 6) { break; }
  }
  return icons;
}

const totalAmenities = ({ basics, facilities, dining, safety }) => {
  return basics.concat(facilities, dining, safety).length;
}

class Amenities extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      modalIsOpen: false
    };
    this.toggleModal = this.toggleModal.bind(this);
  }

  toggleModal() {
    this.setState((prevState) => ({
      modalIsOpen: !prevState.modalIsOpen
    }));
  }

  render() {
    const { amenities } = this.props;
    let icons = assembleIcons(amenities);
    let total = totalAmenities(amenities);
    return(
      <StyledAmenities>
        <div style={title}>Amenities</div>
        <AmenitiesContainer>
          {icons.map((icon, index) => {
            return(
              <AmenityContainer key={index}>
                <AmenityIcon src={icon[1]}></AmenityIcon>
                <span>{capitalize(icon[0])}</span>
              </AmenityContainer>
            )
          })}
        </AmenitiesContainer>
        <ExpansionLink onClick={() => {this.toggleModal()}}>Show all {total} amenities</ExpansionLink>
        <AmenityModal 
          show={this.state.modalIsOpen} 
          amenities={amenities} 
          onClose={this.toggleModal}/>
      </StyledAmenities>
    )
  }
};

Amenities.propTypes = {
  amenities: PropTypes.objectOf(PropTypes.oneOfType([PropTypes.string, PropTypes.array]))
};

Amenities.defaultProps = {
  amenities: {
    basics: [],
    facilities: [],
    dining: [],
    safety: []
  }
};

export default Amenities;