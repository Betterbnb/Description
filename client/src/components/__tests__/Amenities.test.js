import React from 'react';
import { shallow } from 'enzyme';
import Amenities from '../Amenities.jsx';

describe('<Amenities />', () => {
  let wrapper;

  beforeEach(() => {
    const testData = {
      "amenities": {
        "basics": [
          "smoking allowed",
          "wifi",
          "air conditioning",
          "fireplace",
          "TV",
          "essentials (towels, bedsheets, soap, toilet paper, pillows)",
          "shampoo",
          "parking",
          "iron",
          "pets allowed",
          "heating",
          "desk/workspace"
        ],
        "facilities": [
          "washer",
          "pool",
          "gym"
        ],
        "dining": [
          "grill",
          "breakfast",
          "coffee/tea",
          "pizza oven"
        ],
        "safety": "safety card"
      }
    };
    wrapper = shallow(<Amenities {...testData}/>);
  });

  it('should render correctly with props', () => {
    expect(wrapper).toMatchSnapshot();
  });

  xit('should expand options in modal window', () => {
    //write test to open/close modal window
  });

  xit('should scroll with page in modal window', () => {
    //write test to scroll modal window
  });

});