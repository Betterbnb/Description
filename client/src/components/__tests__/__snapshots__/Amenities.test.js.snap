// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Amenities /> should render correctly with props 1`] = `
<Amenities__StyledAmenities>
  <div
    style={
      Object {
        "fontSize": "16px",
        "fontWeight": "500",
        "marginBottom": "15px",
      }
    }
  >
    Amenities
  </div>
  <Amenities__AmenitiesContainer>
    <Amenities__AmenityContainer
      key="0"
    >
      <Amenities__AmenityIcon
        src="https://s3-us-west-1.amazonaws.com/betterbnb-description/wifi.png"
      />
      <span>
        Wifi
      </span>
    </Amenities__AmenityContainer>
    <Amenities__AmenityContainer
      key="1"
    >
      <Amenities__AmenityIcon
        src="https://s3-us-west-1.amazonaws.com/betterbnb-description/fireplace.png"
      />
      <span>
        Fireplace
      </span>
    </Amenities__AmenityContainer>
    <Amenities__AmenityContainer
      key="2"
    >
      <Amenities__AmenityIcon
        src="https://s3-us-west-1.amazonaws.com/betterbnb-description/TV.png"
      />
      <span>
        TV
      </span>
    </Amenities__AmenityContainer>
    <Amenities__AmenityContainer
      key="3"
    >
      <Amenities__AmenityIcon
        src="https://s3-us-west-1.amazonaws.com/betterbnb-description/essentials.png"
      />
      <span>
        Essentials (towels, bedsheets, soap, toilet paper, pillows)
      </span>
    </Amenities__AmenityContainer>
    <Amenities__AmenityContainer
      key="4"
    >
      <Amenities__AmenityIcon
        src="https://s3-us-west-1.amazonaws.com/betterbnb-description/shampoo.png"
      />
      <span>
        Shampoo
      </span>
    </Amenities__AmenityContainer>
    <Amenities__AmenityContainer
      key="5"
    >
      <Amenities__AmenityIcon
        src="https://s3-us-west-1.amazonaws.com/betterbnb-description/parking.png"
      />
      <span>
        Parking
      </span>
    </Amenities__AmenityContainer>
  </Amenities__AmenitiesContainer>
  <Amenities__ExpansionLink
    onClick={[Function]}
  >
    Show all 
    20
     amenities
  </Amenities__ExpansionLink>
  <AmenityModal
    amenities={
      Object {
        "basics": Array [
          "smoking allowed",
          "wifi",
          "air conditioning",
          "fireplace",
          "TV",
          "essentials (towels, bedsheets, soap, toilet paper, pillows)",
          "shampoo",
          "parking",
          "iron",
          "pets allowed",
          "heating",
          "desk/workspace",
        ],
        "dining": Array [
          "grill",
          "breakfast",
          "coffee/tea",
          "pizza oven",
        ],
        "facilities": Array [
          "washer",
          "pool",
          "gym",
        ],
        "safety": "safety card",
      }
    }
    onClose={[Function]}
    show={false}
  />
</Amenities__StyledAmenities>
`;
